jetbrains-core/src/software/aws/toolkits/jetbrains/services/codewhisperer/editor/CodeWhispererTypedHandler.kt

```diff
@@ -4,29 +4,44 @@
 package software.aws.toolkits.jetbrains.services.codewhisperer.editor
 
 import com.intellij.codeInsight.editorActions.TypedHandlerDelegate
+import com.intellij.openapi.application.runInEdt
 import com.intellij.openapi.editor.Editor
 import com.intellij.openapi.project.Project
 import com.intellij.psi.PsiFile
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.isActive
+import kotlinx.coroutines.launch
+import software.aws.toolkits.jetbrains.core.coroutines.projectCoroutineScope
 import software.aws.toolkits.jetbrains.services.codewhisperer.service.CodeWhispererInvocationStatus
 import software.aws.toolkits.jetbrains.services.codewhisperer.service.CodeWhispererService
 import software.aws.toolkits.jetbrains.services.codewhisperer.util.CodeWhispererConstants
 import software.aws.toolkits.telemetry.CodewhispererAutomatedTriggerType
 import software.aws.toolkits.telemetry.CodewhispererTriggerType
 
 class CodeWhispererTypedHandler : TypedHandlerDelegate(), CodeWhispererAutoTriggerHandler {
+    private var triggerOnIdle: Job? = null
     override fun charTyped(c: Char, project: Project, editor: Editor, psiFiles: PsiFile): Result {
+        triggerOnIdle?.cancel()
         if (!CodeWhispererService.getInstance().canDoInvocation(editor, CodewhispererTriggerType.AutoTrigger)) {
             return Result.CONTINUE
         }
         if (CodeWhispererConstants.SPECIAL_CHARACTERS_LIST.contains(c.toString())) {
             performAutomatedTriggerAction(editor, CodewhispererAutomatedTriggerType.SpecialCharacters)
             return Result.CONTINUE
         }
-        val invocationStatus = CodeWhispererInvocationStatus.getInstance()
-        if (invocationStatus.checkKeyStrokeCountMeetThreshold()) {
-            performAutomatedTriggerAction(editor, CodewhispererAutomatedTriggerType.KeyStrokeCount)
-        } else {
-            invocationStatus.incrementKeyStrokeCount()
+        triggerOnIdle = projectCoroutineScope(project).launch {
+            while (!CodeWhispererInvocationStatus.getInstance().hasEnoughDelayToInvokeCodeWhisperer() ||
+                CodeWhispererInvocationStatus.getInstance().hasExistingInvocation()
+            ) {
+                if (!isActive) return@launch
+                delay(CodeWhispererConstants.IDLE_TIME_CHECK_INTERVAL)
+            }
+            runInEdt {
+                if (CodeWhispererInvocationStatus.getInstance().isPopupActive()) return@runInEdt
+                performAutomatedTriggerAction(editor, CodewhispererAutomatedTriggerType.IdleTime)
+                triggerOnIdle?.cancel()
+            }
```

Comment:
There might be a rare race condition issue about when we check `CodeWhispererInvocationStatus.getInstance().hasExistingInvocation()` and when we enter Edt (to invoke CW). However the worst case it can happen is in `performAutomatedTriggerAction` the `hasExistingInvocation` check returns true and returns, so no invocation will happen. It's better we can add a TODO here mentioning this (not a blocker for the PR)

Reply:
The `triggerOnIdle?.cancel()` inside Edt may not be necessary, when it's in Edt the coroutine that schedules it should have finished.

Reply:
updated, thanks!